### PR TITLE
Fix: disallow inf/NaN for Float64 in logical expressions (analyzer)

### PR DESCRIPTION
**Pull request by devcrafter**
Related to https://github.com/ClickHouse/ClickHouse/issues/57324

### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

---
### Modify your CI run:
**NOTE:** If your merge the PR with modified CI you **MUST KNOW** what you are doing
**NOTE:** Set desired options before CI starts or re-push after updates

#### Run only:
- [ ]  Integration tests
- [ ]  Integration tests (arm64)
- [ ]  Stateless tests (release)
- [ ]  Stateless tests (asan)
- [ ]  Stateful tests (release)
- [ ]  Stateful tests (asan)
- [ ]  No sanitizers
- [ ]  Tests with analyzer
- [ ]  Fast tests
- [ ]  Only package_debug build
- [ ]  Add your CI variant description here

#### CI options:
- [ ]  do not test (only style check)
- [ ]  disable merge-commit (no merge from master before tests)
- [ ]  disable CI cache (job reuse)

#### Only specified batches in multi-batch jobs:
- [ ]  1
- [ ]  2
- [ ]  3
- [ ]  4

### CODE REVIEW

**Comment by robot-ch-test-poll4** [comment]
*This is an automated comment for commit 77ddeb9cd56e8553158fa9642a3eca13cea59830 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/61958/77ddeb9cd56e8553158fa9642a3eca13cea59830/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

**Review by KochetovNicolai — APPROVED** [review]

**Comment by devcrafter** [comment]
https://s3.amazonaws.com/clickhouse-test-reports/61958/05260fd0518adc7a4abc2431eb6389c3485121e0/ast_fuzzer__asan_.html - should be fixed by https://github.com/ClickHouse/ClickHouse/pull/62151

File: tests/queries/0_stateless/01386_negative_float_constant_key_condition.sql [hunk 1]
```diff
@@ -15,7 +15,7 @@ INSERT INTO t0 VALUES (0, 0);
 
 SELECT t0.c1 FROM t0 WHERE NOT (t0.c1 OR (t0.c0 AND -1524532316));
 SELECT t0.c1 FROM t0 WHERE NOT (t0.c1 OR (t0.c0 AND -1.0)); -- { serverError 70 }
-SELECT t0.c1 FROM t0 WHERE NOT (t0.c1 OR (t0.c0 AND inf));
-SELECT t0.c1 FROM t0 WHERE NOT (t0.c1 OR (t0.c0 AND nan));
+SELECT t0.c1 FROM t0 WHERE NOT (t0.c1 OR (t0.c0 AND inf)); -- { serverError ILLEGAL_TYPE_OF_ARGUMENT }
```
Comment (devcrafter) [inline tests/queries/0_stateless/01386_negative_float_constant_key_condition.sql:18]: @alexey-milovidov It seems you've added `inf` and `nan` here intentionally some time ago. Could you clarify the motivation behind it? It looks at least controversial

Comment (alexey-milovidov) [reply tests/queries/0_stateless/01386_negative_float_constant_key_condition.sql:18]: The motivation is the following:
1. Not only UInt8 is supported in the arguments of logical operators, but other numeric types as well.
2. It is reasonable for all numeric types which include 0 and 1.
3. The behavior should be the same as in C/C++: when the value is non-zero, it behaves as true.
4. That explains why it should work for Floating types as well.
5. The behavior for `inf` and `nan` does not have to be as well defined, but we should have at least one test to assert it.
6. That's why the text exists.

Comment (alexey-milovidov) [reply tests/queries/0_stateless/01386_negative_float_constant_key_condition.sql:18]: 7. I don't mind if you change it, and add a note to the changelog.

**Pull request closed without merging**